src/neuroglancer/sliceview/compresso/compresso.hpp

```diff
@@ -0,0 +1,593 @@
+					}
+					labels[loc] = labels[loc - sx];
+				}
+				else if (offset == 3) {
+					if (y >= sy - 1) {
+						return 5;
+					}
+					labels[loc] = labels[loc + sx];
+				}
+				else if (offset == 4) {
+					if (z == 0) {
+						return 6;
+					}
+					labels[loc] = labels[loc - sxy];
+				}
+				else if (offset == 5) {
+					if (z >= sz - 1) {
+						return 7;
+					}
+					labels[loc] = labels[loc + sxy];
+				}
+				else if (offset == 6) {
+					labels[loc] = locations[index + 1];
+					index++;
+				}
+				else {
+					labels[loc] = offset - 7;
+				}
+				index++;
+			}
+		}
+	}
+
+	return 0;
+}
+
+template <typename LABEL, typename WINDOW>
+int decompress(unsigned char* buffer, size_t num_bytes, LABEL* output = NULL) {
+
+	if (num_bytes < CompressoHeader::header_size) {
+		return 8;
+	}
+	else if (!CompressoHeader::valid_header(buffer)) {
+		return 9;
+	}
+
+	const CompressoHeader header(buffer);
+
+	const size_t sx = header.sx;
+	const size_t sy = header.sy;
+	const size_t sz = header.sz;
+	const size_t voxels = sx * sy * sz;
+	const size_t xstep = header.xstep;
+	const size_t ystep = header.ystep;
+	const size_t zstep = header.zstep;
+
+	if (sx * sy * sz == 0) {
+		return 10;
+	}
+
+	const size_t nx = (sx + xstep - 1) / xstep; // round up
+	const size_t ny = (sy + ystep - 1) / ystep; // round up
+	const size_t nz = (sz + zstep - 1) / zstep; // round up
+	const size_t nblocks = nz * ny * nx;
+
+	size_t window_bytes = (
+		num_bytes 
+			- CompressoHeader::header_size
+			- (header.id_size * sizeof(LABEL))  
+			- (header.value_size * sizeof(WINDOW))
+			- (header.location_size * sizeof(LABEL))
+	);
+	size_t num_condensed_windows = window_bytes / sizeof(WINDOW);
+
+	// allocate memory for all arrays
+	std::vector<LABEL> ids(header.id_size + 1); // +1 to allow vectorized mapping w/ no if statement guarding zero
+	std::vector<WINDOW> window_values(header.value_size);
+	std::vector<LABEL> locations(header.location_size);
+	std::vector<WINDOW> windows(num_condensed_windows);
+
+	size_t iv = CompressoHeader::header_size;
+	for (size_t ix = 0; ix < ids.size() - 1; ix++, iv += sizeof(LABEL)) {
+		ids[ix + 1] = ctoi<LABEL>(buffer, iv);
+	}
+	for (size_t ix = 0; ix < window_values.size(); ix++, iv += sizeof(WINDOW)) {
+		window_values[ix] = ctoi<WINDOW>(buffer, iv);
+	}
+	for (size_t ix = 0; ix < locations.size(); ix++, iv += sizeof(LABEL)) {
+		locations[ix] = ctoi<LABEL>(buffer, iv);
+	}
+	for (size_t ix = 0; ix < num_condensed_windows; ix++, iv += sizeof(WINDOW)) {
+		windows[ix] = ctoi<WINDOW>(buffer, iv);
+	}
+
+	windows = run_length_decode_windows<WINDOW>(windows, nblocks);
+
+	bool* boundaries = decode_boundaries<WINDOW>(
+		windows, window_values, 
+		sx, sy, sz, 
+		xstep, ystep, zstep
+	);
+	windows = std::vector<WINDOW>();
+	window_values = std::vector<WINDOW>();
+
+	uint32_t* components = cc3d::connected_components<uint32_t>(
+		boundaries, sx, sy, sz, header.connectivity
+	);
+
+	if (output == NULL) {
+		return 11;
+	}
+
```

Comment:
What is this condition for?

Reply:
Just in case someone passes in NULL by accident. In the cpp library edition, it will create a new array and return it instead. The position can be moved up.